File: AtCoder/ABC/158/python/D_String_Formation.py
S = input()
Q = int(input())
f = False
front, back = [], []
for _ in range(Q):
    query = input().split()
    if len(query) == 1:
        f = not f
    else:
        F, C = query[1:]
        if F == "1":
            if f:
                back.append(C)
            else:
                front.append(C)
        else:
            if f:
                front.append(C)
            else:
                back.append(C)
if f:
    res = "".join(back[::-1]) + S[::-1] + "".join(front)
else:
    res = "".join(front[::-1]) + S + "".join(back)
print(res)

File: AtCoder/ABC/158/cpp/D_String_Formation.cpp
#include <bits/stdc++.h>
using namespace std;

string S;
int Q, T, F;
char C;

int main() {
  cin >> S >> Q;
  bool f = false;
  string front, back;
  vector<bool> v;
  for (int i = 0; i < Q; ++i) {
    cin >> T;
    v.push_back(f);
    if (T == 1) {
      f = !f;
    } else {
      cin >> F >> C;
      if (F == 1) {
        if (f) {
          back.push_back(C);
        } else {
          front.push_back(C);
        }
      } else {
        if (f) {
          front.push_back(C);
        } else {
          back.push_back(C);
        }
      }
    }
  }
  string res;
  if (f) {
    reverse(back.begin(), back.end());
    reverse(S.begin(), S.end());
    res = back + S + front;
  } else {
    reverse(front.begin(), front.end());
    res = front + S + back;
  }
  puts(res.c_str());
}